src/Items/ItemDye.h

```diff
@@ -216,51 +216,185 @@ class cItemDyeHandler :
+					case 6: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_WHITE_TULIP); break;
+					case 7: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_ORANGE_TULIP); break;
+					case 8: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_OXEYE_DAISY); break;
+					// TODO: Add cornflower
+				}
+				break;
+			}
+			case biSwampland:
+			case biSwamplandM:
+			{
+				a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_BLUE_ORCHID);
+				break;
+			}
+			case biFlowerForest:
+			{
+				switch (Random.RandInt(8))
+				{
+					case 0: a_World.SetBlock(a_Position, E_BLOCK_DANDELION, 0); break;
+					case 1: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_POPPY); break;
+					case 2: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_ALLIUM); break;
+					case 3: a_World.SetBlock(a_Position, E_BLOCK_RED_ROSE, 0); break;  // was renamed to Azure Bluet later
+					case 4: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_RED_TULIP); break;
+					case 5: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_PINK_TULIP); break;
+					case 6: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_WHITE_TULIP); break;
+					case 7: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_ORANGE_TULIP); break;
+					case 8: a_World.SetBlock(a_Position, E_BLOCK_FLOWER, E_META_FLOWER_OXEYE_DAISY); break;
+					// TODO: Add cornflower, lily of the valley
+				}
+				break;
+			}
+			case biMesa:
+			case biMesaBryce:
+			case biMesaPlateau:
+			case biMesaPlateauF:
+			case biMesaPlateauM:
+			case biMesaPlateauFM:
+			case biMushroomIsland:
+			case biMushroomShore:
+			case biNether:
+			case biEnd:
+			{
+				break;
+			}
+			default:
+			{
+				switch (Random.RandInt(1))
+				{
+					case 0: a_World.SetBlock(a_Position, E_BLOCK_DANDELION, 0); break;
+					case 1: a_World.SetBlock(a_Position, E_BLOCK_RED_ROSE, 0); break;
+				}
+				break;
+			}
+		}
+	}
+
+	/** Walks adjacent grass blocks up to 7 taxicab distance away from a_Position and calls the Planter function on the first suitable one found.
+	Does nothing if no position suitable for growing was found. */
+	template <auto Planter>
+	static void FindAdjacentGrassAnd(cWorld & a_World, const Vector3i a_Position)
+	{
+		auto & Random = GetRandomProvider();
+		auto Position = a_Position;
+
+		// Maximum 7 taxicab distance away from centre:
+		for (
+			int Tries = 0;
+			Tries != 8;
+			Tries++,
+
+			// Get the adjacent block to visit this iteration:
+			Position += Vector3i(
+				Random.RandInt(-1, 1),
+				Random.RandInt(-1, 1) * (Random.RandInt(2) / 2),  // Y offset, with discouragement to values that aren't zero
+				Random.RandInt(-1, 1)
+			)
+		)
+		{
+			if (
+				!cChunkDef::IsValidHeight(Position.y) ||
+				(a_World.GetBlock(Position) != E_BLOCK_GRASS)  // Are we looking at grass?
+			)
+			{
+				// Not grass or invalid height, restart random walk and bail:
+				Position = a_Position;
```

Comment:
Maybe also reset the the Tries variable so that we completely restart the tun

Reply:
Although that might result in infinite loops when there's actually no space for a plant

Reply:
Yes, Tries acts a sort of a "fuel limit" for the loop as well as a walk distance limit